File: modules/face_detector.py
import cv2
import mediapipe as mp
import numpy as np
import time
from mediapipe.tasks import python
from mediapipe.tasks.python import vision
from core.base_module import BaseModule
from utils.logger import logger

class FaceDetectorModule(BaseModule):
    """
    Module 1: Face Detection + Tracking
    Uses MediaPipe Tasks for high-performance multi-face detection.
    Includes persistent ID tracking logic with temporal memory.
    """

    def __init__(self, name: str, config: dict = None):
        super().__init__(name, config)

        # Initialize MediaPipe Face Detector Task
        base_options = python.BaseOptions(model_asset_path='models/face_detector.tflite')
        options = vision.FaceDetectorOptions(
            base_options=base_options,
            min_detection_confidence=self.config.get('min_detection_confidence', 0.5)
        )
        self.detector = vision.FaceDetector.create_from_options(options)

        # Tracking state
        self.next_id = 0
        # {id: {'centroid': (x,y), 'last_seen': timestamp, 'disappeared': frames}}
        self.tracked_faces = {}
        self.max_disappeared = 15 # Frames to keep ID if face is lost
        logger.info("Face Detector Module initialized with MediaPipe Tasks and Persistent Tracking.")

    def _calculate_distance(self, c1, c2):
        return np.sqrt((c1[0] - c2[0])**2 + (c1[1] - c2[1])**2)

    def process(self, frame: np.ndarray) -> dict:
        """
        Detect faces and extract bounding boxes.
        Includes Centroid Tracking with frame persistence.
        """
        results = {}
        rgb_frame = cv2.cvtColor(frame, cv2.COLOR_BGR2RGB)
        mp_image = mp.Image(image_format=mp.ImageFormat.SRGB, data=rgb_frame)

        detection_result = self.detector.detect(mp_image)

        current_detections = []
        if detection_result.detections:
            for detection in detection_result.detections:
                bbox = detection.bounding_box
                x, y, w, h = bbox.origin_x, bbox.origin_y, bbox.width, bbox.height
                score = detection.categories[0].score
                centroid = (x + w // 2, y + h // 2)

                current_detections.append({
                    'bbox': [x, y, w, h],
                    'confidence': score,
                    'centroid': centroid
                })

        # --- Persistent Centroid Tracking ---
        # 1. Update existing tracks
        matched_detections = set()
        updated_tracks = {}

        # Copy current tracks to attempt matching
        track_ids = list(self.tracked_faces.keys())

        if len(current_detections) > 0:
            for tid in track_ids:
                tdata = self.tracked_faces[tid]

                # Find closest detection
                min_dist = 50 # threshold
                best_idx = -1

                for i, det in enumerate(current_detections):
                    if i in matched_detections: continue
                    dist = self._calculate_distance(tdata['centroid'], det['centroid'])
                    if dist < min_dist:
                        min_dist = dist
                        best_idx = i

                if best_idx != -1:
                    # Match found
                    det = current_detections[best_idx]
                    det['id'] = tid
                    updated_tracks[tid] = {
                        'centroid': det['centroid'],
                        'disappeared': 0
                    }
                    matched_detections.add(best_idx)
                else:
                    # No match for this track
                    tdata['disappeared'] += 1
                    if tdata['disappeared'] <= self.max_disappeared:
                        updated_tracks[tid] = tdata
        else:
            # No detections, increment all disappeared
            for tid in track_ids:
                self.tracked_faces[tid]['disappeared'] += 1
                if self.tracked_faces[tid]['disappeared'] <= self.max_disappeared:
                    updated_tracks[tid] = self.tracked_faces[tid]

        # 2. Register new detections
        for i, det in enumerate(current_detections):
            if i not in matched_detections:
                det['id'] = self.next_id
                updated_tracks[self.next_id] = {
                    'centroid': det['centroid'],
                    'disappeared': 0
                }
                self.next_id += 1

        self.tracked_faces = updated_tracks
        results['faces'] = [d for d in current_detections if 'id' in d]
        return results

    def draw(self, frame: np.ndarray, results: dict) -> np.ndarray:
        faces = results.get('faces', [])
        for face in faces:
            x, y, w, h = face['bbox']
            conf = face['confidence']
            face_id = face.get('id', '?')

            cv2.rectangle(frame, (x, y), (x + w, y + h), (0, 255, 0), 2)
            label = f"ID: {face_id} [{conf:.2f}]"
            cv2.putText(frame, label, (x, y - 10),
                        cv2.FONT_HERSHEY_SIMPLEX, 0.5, (0, 255, 0), 2)
        return frame

File: js/modules/FaceDetector.js
/**
 * Module 1: Face Detection + Tracking (Client-side)
 * Uses MediaPipe Face Detection JS SDK.
 */
class FaceDetectorModule extends BaseModule {
    constructor(name) {
        super(name);
        this.faceDetection = null;
        this.results = null;

        // Tracking state
        this.nextId = 0;
        this.trackedFaces = new Map(); // id -> {centroid: {x,y}, disappeared: frames}
        this.maxDisappeared = 20;
    }

    async init() {
        this.faceDetection = new FaceDetection({
            locateFile: (file) => {
                return `https://cdn.jsdelivr.net/npm/@mediapipe/face_detection/${file}`;
            }
        });

        this.faceDetection.setOptions({
            model: 'short',
            minDetectionConfidence: 0.5
        });

        this.faceDetection.onResults((results) => {
            this.results = results;
        });

        console.log("FaceDetectorModule initialized.");
    }

    async process(videoElement) {
        await this.faceDetection.send({ image: videoElement });

        if (!this.results || !this.results.detections) return { faces: [] };

        const currentDetections = this.results.detections.map(det => {
            const bbox = det.boundingBox;
            return {
                bbox: [bbox.xCenter - bbox.width/2, bbox.yCenter - bbox.height/2, bbox.width, bbox.height],
                centroid: { x: bbox.xCenter, y: bbox.yCenter },
                confidence: det.categories[0].score
            };
        });

        // Persistent Tracking Logic
        const updatedTracks = new Map();
        const matchedIndices = new Set();

        // 1. Update existing tracks
        for (const [id, tdata] of this.trackedFaces.entries()) {
            let minDist = 0.08; // threshold
            let bestIdx = -1;

            for (let i = 0; i < currentDetections.length; i++) {
                if (matchedIndices.has(i)) continue;
                const dist = Math.sqrt(
                    Math.pow(tdata.centroid.x - currentDetections[i].centroid.x, 2) +
                    Math.pow(tdata.centroid.y - currentDetections[i].centroid.y, 2)
                );
                if (dist < minDist) {
                    minDist = dist;
                    bestIdx = i;
                }
            }

            if (bestIdx !== -1) {
                currentDetections[bestIdx].id = id;
                updatedTracks.set(id, { centroid: currentDetections[bestIdx].centroid, disappeared: 0 });
                matchedIndices.add(bestIdx);
            } else {
                tdata.disappeared++;
                if (tdata.disappeared <= this.maxDisappeared) {
                    updatedTracks.set(id, tdata);
                }
            }
        }

        // 2. Add new detections
        for (let i = 0; i < currentDetections.length; i++) {
            if (!matchedIndices.has(i)) {
                currentDetections[i].id = this.nextId;
                updatedTracks.set(this.nextId, { centroid: currentDetections[i].centroid, disappeared: 0 });
                this.nextId++;
            }
        }

        this.trackedFaces = updatedTracks;
        return { faces: currentDetections.filter(d => d.id !== undefined) };
    }

    draw(ctx, results, canvas) {
        if (!results || !results.faces) return;

        const { width, height } = canvas;

        results.faces.forEach(face => {
            const [x, y, w, h] = face.bbox;

            // Draw Box
            ctx.strokeStyle = '#00ff88';
            ctx.lineWidth = 3;
            ctx.strokeRect(x * width, y * height, w * width, h * height);

            // Draw Label
            ctx.fillStyle = '#00ff88';
            ctx.font = 'bold 16px Courier New';
            const label = `ID:${face.id} [${(face.confidence * 100).toFixed(0)}%]`;
            ctx.fillText(label, x * width, y * height - 10);

            // Draw scan lines effect
            ctx.fillStyle = 'rgba(0, 255, 136, 0.1)';
            ctx.fillRect(x * width, y * height, w * width, 2);
        });
    }
}
